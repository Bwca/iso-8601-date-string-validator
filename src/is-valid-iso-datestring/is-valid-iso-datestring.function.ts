import { getStringSeparator } from '../get-string-separator/get-string-separator.function';
import { isValidDate } from '../is-valid-date/is-valid-date.function';
import { isValidTime } from '../is-valid-time/is-valid-time.function';

import { getTimeStringSeparator } from './get-time-string-separator/get-time-string-separator.function';

export function isValidISODateString(dateString: string): boolean {
  const [date, timeWithOffset] = dateString.split('T');
  const dateSeparator = getStringSeparator(date);
  const isDateValid = isValidDate(date, dateSeparator);

  if (!timeWithOffset) {
    return false;
  }

  const timeStringSeparator = getTimeStringSeparator(timeWithOffset);

  return isDateValid && isValidTime(timeWithOffset, timeStringSeparator, true);
}
