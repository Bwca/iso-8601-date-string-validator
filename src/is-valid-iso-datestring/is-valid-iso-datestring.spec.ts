import { SOME_VALID_ISO_DATETIME_STRINGS } from '../tests/constants/some-valid-iso-strings.const';
import { NEGATIVE_TIMEZONES } from '../tests/constants/timezones/negative-timezones.const';
import { POSITIVE_TIMEZONES } from '../tests/constants/timezones/positive-timezones.const';

import { isValidISODateString } from './is-valid-iso-datestring.function';

const noZoneValidISODateTimeString = '2019-07-09T15:03:36';

test(`isValidISODateString. expect all valid ISO 8601 strings to validate`, () => {
  // Arrange
  let results: boolean[];

  // Act
  results = SOME_VALID_ISO_DATETIME_STRINGS.map((s) => isValidISODateString(s));

  // Assert
  expect(results).not.toContain(false);
  expect(results).toMatchSnapshot();
});

test(`isValidISODateString. expect a valid ISO string to validate with all negative zone offsets`, () => {
  // Arrange
  let results: boolean[];

  // Act
  results = NEGATIVE_TIMEZONES.map((s) =>
    isValidISODateString(`${noZoneValidISODateTimeString}-${s}`),
  );

  // Assert
  expect(results).not.toContain(false);
  expect(results).toMatchSnapshot();
});

test(`isValidISODateString. expect a valid ISO string to validate with all positive zone offsets`, () => {
  // Arrange
  let results: boolean[];

  // Act
  results = POSITIVE_TIMEZONES.map((s) =>
    isValidISODateString(`${noZoneValidISODateTimeString}+${s}`),
  );

  // Assert
  expect(results).not.toContain(false);
  expect(results).toMatchSnapshot();
});

test('isValidISODateString does not validate date without time', () => {
  // Arrange
  const date = '2020-12-04';

  // Act
  const isValid = isValidISODateString(date);

  // Assert
  expect(isValid).toBeFalsy();
});
