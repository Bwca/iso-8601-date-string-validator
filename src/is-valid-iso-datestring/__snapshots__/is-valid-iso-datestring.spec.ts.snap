// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core isValidISODateString tests isValidISODateString. expect a valid ISO string to validate with all negative zone offsets 1`] = `
Array [
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
]
`;

exports[`Core isValidISODateString tests isValidISODateString. expect a valid ISO string to validate with all positive zone offsets 1`] = `
Array [
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
  true,
]
`;

exports[`Core isValidISODateString tests isValidISODateString. expect all valid ISO 8601 strings to validate 1`] = `
Array [
  true,
  true,
  true,
  true,
  true,
]
`;
